rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/dap/breakpoint/BreakpointsCollection.java

```diff
@@ -0,0 +1,45 @@
+package org.rascalmpl.vscode.lsp.dap.breakpoint;
+
+import io.usethesource.vallang.ISourceLocation;
+import org.eclipse.lsp4j.debug.Source;
+import org.rascalmpl.debug.DebugHandler;
+import org.rascalmpl.debug.DebugMessageFactory;
+
+import java.util.HashMap;
+
+/**
+    Store breakpoints set by user
+ **/
+
+public class BreakpointsCollection {
+    private final HashMap<String, HashMap<ISourceLocation, BreakpointInfo>> breakpoints;
+    private final DebugHandler debugHandler;
+    private int breakpointIDCounter = 0;
+
+    public BreakpointsCollection(DebugHandler debugHandler){
+        this.breakpoints = new HashMap<>();
+        this.debugHandler = debugHandler;
+    }
+
+    public void clearBreakpointsOfFile(String filePath){
+        if(!breakpoints.containsKey(filePath)) return;
+        for (ISourceLocation breakpointLocation : breakpoints.get(filePath).keySet()) {
+            debugHandler.processMessage(DebugMessageFactory.requestDeleteBreakpoint(breakpointLocation));
+        }
+        breakpoints.get(filePath).clear();
+    }
+
+    public void addBreakpoint(ISourceLocation location, Source source){
+        String path = location.getPath();
+        if(!breakpoints.containsKey(path)) breakpoints.put(path, new HashMap<>());
+        BreakpointInfo breakpoint = new BreakpointInfo(++breakpointIDCounter, source);
+        breakpoints.get(path).put(location, breakpoint);
+        debugHandler.processMessage(DebugMessageFactory.requestSetBreakpoint(location));
+    }
+
+    public int getBreakpointID(ISourceLocation location){
+        String path = location.getPath();
+        if(!breakpoints.containsKey(path) || !breakpoints.get(path).containsKey(location)) return -1;
```

Comment:
similar, better to do a single get. now this code has 3 lookups in the breakpoints map, and 2 in the nested map.